File: Gnome_sort/Python/jcla1/gnome_sort.py
# Gnome sort runs in O(n^2), but if the array
# is nearly sorted already it runs in about O(n)
# http://en.wikipedia.org/wiki/Gnome_sort
def gnome_sort(arr):
    pos = 1
    while pos < len(arr):
        if arr[pos] >= arr[pos-1]:
            pos += 1
        else:
            arr[pos], arr[pos-1] = arr[pos-1], arr[pos]
            if pos > 1:
                pos -= 1
    return arr

File: Gnome_sort/Lua/Yonaba/gnome_sort.lua
-- Gnome sort implementation
-- See: http://en.wikipedia.org/wiki/Gnome_sort

-- array   : an array to be sorted, in-place
-- comp    : (Optional) a comparison function
--  defaults to function(a,b) return a < b end
-- returns : the passed-in array, sorted
return function (array, comp)
  comp = comp or function(a,b) return a < b end
  local n = #array
  local pos = 2
  while pos <= n do
    if comp(array[pos-1], array[pos]) then
      pos = pos + 1
    else
      array[pos-1], array[pos] = array[pos], array[pos-1]
      if pos > 2 then pos = pos - 1 end
    end
  end
  return array
end
